src/net.cpp

```diff
@@ -950,14 +963,18 @@ void ThreadSocketHandler()
             if (FD_ISSET(pnode->hSocket, &fdsetRecv) || FD_ISSET(pnode->hSocket, &fdsetError))
             {
                 TRY_LOCK(pnode->cs_vRecvMsg, lockRecv);
-                if (lockRecv)
+                int64_t amt2Recv = receiveShaper.available(RECV_SHAPER_MIN_FRAG);
+                if (lockRecv&&(amt2Recv>0))
                 {
                     {
-                        // typical socket buffer is 8K-64K
-                        char pchBuf[0x10000];
+                        progress++;
+                        // max of min makes sure amt is in a range reasonable for buffer allocation
+                        int64_t amt = max((int64_t)1,min(amt2Recv,MAX_RECV_CHUNK));
+                        char pchBuf[amt];
```

Comment:
I worry a bit about this being on the stack. It would be easy for it to accidentally end up too large, or having weird platform dependencies on operating systems that default to tiny stacks.

It might be better to move this onto the heap and access it via pnode instead.